### PR TITLE
kubernetes cluster networking types based on IP address assignment

### PR DESCRIPTION
**Pull request by aojea**
Kubernetes has a concept of cluster IP family, as it can be IPv4 only, IPv6 only or dual stack depending on the IP famliies configured for their Pod , Services and Node objects.

### CODE REVIEW

File: content/en/docs/concepts/cluster-administration/networking.md
Comment (sftim) [inline content/en/docs/concepts/cluster-administration/networking.md:37]: (nit)
```suggestion
## Cluster networking types {#cluster-network-ipfamilies}
```

File: content/en/docs/concepts/cluster-administration/networking.md
Comment (sftim) [inline content/en/docs/concepts/cluster-administration/networking.md:47]: (nit)
```suggestion
independently of the existing IPs of the represented objects. Per example, a server or a pod can have multiple
IP addresses on its interfaces, but only the IP addresses in `node.status.addresses` or `pod.status.ips` are
```

**Review by sftim — COMMENTED** [review]
Thanks.

Two nits;
/lgtm

We can merge this as-is

File: content/en/docs/concepts/cluster-administration/networking.md
Comment (danwinship) [inline content/en/docs/concepts/cluster-administration/networking.md:43]: So what is this addition for? Because it's not actually true; the network plugin's configuration determines whether pods are ipv4/ipv6/dual, kube-apiserver's confiugration determines whether services are ipv4/ipv6/dual, and the cloud provider's or kubelet's configuration decides whether nodes are ipv4/ipv6 or dual. Nothing requires them all to be configured the same way...

File: content/en/docs/concepts/cluster-administration/networking.md
Comment (aojea) [inline content/en/docs/concepts/cluster-administration/networking.md:43]:
```suggestion
- IPv4/IPv6 or IPv6/IPv4 dual-stack, Pod and Nodes contain IPv4 and IPv6 addresses and Services may contain IPv4 and/or IPv6 addresses. For dual-stack clusters, the  components must agree on the IP family order to meet the networking model and the first IP family is considered as the preferred one for Services.
```

File: content/en/docs/concepts/cluster-administration/networking.md
Comment (aojea) [reply content/en/docs/concepts/cluster-administration/networking.md]: we have e2e tests that for the order of the IPs and the preferred IP family to verify that clusters meet our definition of network model

Comment (danwinship) [reply content/en/docs/concepts/cluster-administration/networking.md]: > same as nothing requires pods and nodes to be routable, is our definition of network model what requires them to be the same

Historically, we have jumped through hoops to _not_ require that... AFAIK everything works "as expected" when you have inconsistent IP-stack-ness. (If you have dual-stack pods and single-stack IPv4 services, then you just can't create services that point at IPv6 pod IPs. If you have single-stack IPv4 pods and dual-stack services, then your dual-stack services will just never have any IPv6 endpoints. If you have dual-stack pods and dual-stack services and single-stack nodes, then you can have dual-stack pod-network services but not dual-stack host-network services. Etc.)

I'd certainly be happy to say that kubernetes doesn't support clusters with inconsistent IP-stack-ness, but I feel like that's a _change_ in our position, not a clarification of it.

Comment (danwinship) [reply content/en/docs/concepts/cluster-administration/networking.md]: > I'd certainly be happy to say that kubernetes doesn't support clusters with inconsistent IP-stack-ness

Actually, I wouldn't, since that would imply you can't live-migrate a cluster from single-stack to dual-stack.

Comment (aojea) [reply content/en/docs/concepts/cluster-administration/networking.md]: >  AFAIK everything works "as expected" when you have inconsistent IP-stack-ness

I disagree, the cluster will not be functional, that the network works or that the system allows you to do certain operation does not mean the cluster works correctly: controllers,  Incluser authentication , webhooks, DNS, and critical functionalities will not work  ... the network in kubernetes is more than the IPs , we are responsible of anything in the cluster that uses the network

> but I feel like that's a change in our position

is not a change in our position, we have e2e tests asserting this https://github.com/kubernetes/kubernetes/blob/master/test/e2e/network/dual_stack.go , once we define an e2e we define a behavior

Comment (aojea) [reply content/en/docs/concepts/cluster-administration/networking.md]: basically, a cluster with the wrong configuration will not even pass the conformance tests, means is not supported

Comment (aojea) [reply content/en/docs/concepts/cluster-administration/networking.md]: is that fair?

Comment (danwinship) [reply content/en/docs/concepts/cluster-administration/networking.md]: > the cluster will not be functional

What, precisely, would not function? Yes, you can't have a dual-stack service pointing to a single-stack pod, but you can't have a dual-stack service pointing to a single-stack external host either. That's not "not functioning", that's just "you asked it to do something that doesn't make sense".

> we have e2e tests asserting this

We have e2e tests asserting that if you have a fully-dual-stack cluster, that things work in a certain way. I will grant you that we do not have e2e tests for partially-dual-stack clusters.

> a cluster with the wrong configuration will not even pass the conformance tests

which _conformance_ tests will it fail? Nothing in `test/e2e/network/dual_stack.go` is conformance.

Comment (aojea) [reply content/en/docs/concepts/cluster-administration/networking.md]: > since that would imply you can't live-migrate a cluster from single-stack to dual-stack.

temporary states are unrelated, this is an architectural choice not a hard rule, a cluster transitioning states, upgrading or migration does not apply

Comment (danwinship) [reply content/en/docs/concepts/cluster-administration/networking.md]: Also, you cannot plausibly argue that we don't support running a single-stack IPv4 cluster on nodes that happen to also have IPv6 addresses.

Comment (aojea) [reply content/en/docs/concepts/cluster-administration/networking.md]: > which conformance tests will it fail? Nothing in test/e2e/network/dual_stack.go is conformance.

fair, all the e2e tests are part of the kubernetes definition, a subset of them are used by CNCF, but all those tests are defined by the SIGS and eventually can be promoted, dual stack is not mandatory so it will not promoted to CNCF conformance, but we added to establish a standardization and that distros and users and develops can have an homogeneus experience

> the cluster will not be functional

> If you have dual-stack pods and dual-stack services and single-stack nodes, then you can have dual-stack pod-network services but not dual-stack host-network services

ok, I overgeneralized, sorry, there are permutations that as you are pointing out will work, but do we want to support them? does it make sense? why we should not define the correct way to do it to avoid surprises or edge cases that work in some places and not in other? one of the success of kubernetes is that all flavors are "compatible"

With dual stack there are some combinations that will not work, per example if the apiserver primary family and the Node or Pod single stack family does not match, InCluster config will not work., or kubectl exec ...

Anyway, for me the important is the definition of IPv4 only and IPv6 only, If the dual stack point is contentious please let me know how to rephrase, or I just skip it and link to the dual stack existing docs https://kubernetes.io/docs/concepts/services-networking/dual-stack/

Comment (aojea) [reply content/en/docs/concepts/cluster-administration/networking.md]: ok, changed , PTAL , now there is no possible ambiguity,

Comment (aojea) [reply content/en/docs/concepts/cluster-administration/networking.md]: > Also, you cannot plausibly argue that we don't support running a single-stack IPv4 cluster on nodes that happen to also have IPv6 addresses.

This is user facing documentation, not developer facint, I just try to make this simple to understand because when I talk with users and customers a lot of them are confused about this topic,  and only power users or developers will create complex combinations of networking parameters, most people use whatever Openshift, GKE, EKS, AKS, ranchers provides ...

I don't think is worth our time to keep discussing this, if you prefer to not merge this I'm ok with that too, or if you think some part has to be reworded please suggest it

File: content/en/docs/concepts/cluster-administration/networking.md
Comment (sftim) [inline content/en/docs/concepts/cluster-administration/networking.md:43]:
```suggestion
- IPv4/IPv6 or IPv6/IPv4 [dual-stack](/docs/concepts/services-networking/dual-stack/).
```

File: content/en/docs/concepts/cluster-administration/networking.md
Comment (sftim) [reply content/en/docs/concepts/cluster-administration/networking.md]: > it's not actually true; the network plugin's configuration determines whether pods are ipv4/ipv6/dual, kube-apiserver's confiugration determines whether services are ipv4/ipv6/dual, and the cloud provider's or kubelet's configuration decides whether nodes are ipv4/ipv6 or dual. Nothing requires them all to be configured the same way...

Let's be wary of merging anything that can mislead. On the other hand, if Kubernetes can be set up in a way that doesn't pass conformance tests, and the proposed docs are correct for all clusters that pass our tests, I think they're fine to merge.

Which applies here?

**Review by sftim — COMMENTED** [review]
One bug: https://github.com/kubernetes/website/pull/43964#pullrequestreview-1736719746

File: content/en/docs/concepts/cluster-administration/networking.md
Comment (danwinship) [reply content/en/docs/concepts/cluster-administration/networking.md]: > ok, changed , PTAL , now there is no possible ambiguity,

/lgtm

> Let's be wary of merging anything that can mislead.

The latest version doesn't claim that "all components have to be in sync" like the original did. It just doesn't discuss the case where they don't, which I think is fine.

Comment (aojea) [reply content/en/docs/concepts/cluster-administration/networking.md]: > if Kubernetes can be set up in a way that doesn't pass conformance tests,

[I overgeneralized and Dan was right there that there are some combinations that may work ](https://github.com/kubernetes/website/pull/43964#discussion_r1396407933 ), at least I can't proof  which of all the permutations in a dual stack cluster will work, 

I know for sure that there are some combinations that can make conformance test to not pass because while working on dual stack implementations in kubeadm, kind and other installers I hit bugs on `kubectl exec` , webhooks, admission controllers that depends on the Node, Pods, Services objects IPs for the discovery and if they don't work if the order does not match in a dual stack clusters, and those functionalities are exercised in the Conformance tests.

But as I said,  I just wrote how dual-stack is implemented in most of the installers of Kubernetes I know of,

**Comment by aojea** [comment]
@sftim I synced with @danwinship offline, and we thought it is better/easier to understand to make the distinction of the IP families based on the configuration of the components.

Added a new section first to explain the different components and the IP address ranges configured each of them

PTAL and add suggestions if you feel. there are parts that are still not clear or may be misleading

File: content/en/docs/concepts/cluster-administration/networking.md
Comment (danwinship) [inline content/en/docs/concepts/cluster-administration/networking.md:40]: Hm...
- 74 `network plugin`
- 73 `CNI plugin`
- 16 `CNI provider`
- 12 `network add-on`
- 9 networking plugin
- 5 `Container Network Interface (CNI) plugin`
- 3 `Container Network Interface (CNI) provider`
- 1 network plug-in
- 1 networking addon

(case-insensitive, `content/en/` only, not including this PR)
(It's hard to count instances of using "CNI" as a noun, but those exist too.)

Bleah. Anyway, not "networking addon".

Comment (aojea) [reply content/en/docs/concepts/cluster-administration/networking.md:40]: I know, I was going to use "network plugin" but then I saw Line73 and preferred to keep it consistent, but I really don't have a strong opinion, if I have to choose I will go with "network plugin"

> for a non-exhaustive list of networking addons supported by Kubernetes.

Comment (sftim) [reply content/en/docs/concepts/cluster-administration/networking.md:40]: Supporting anecdotal data:

https://kubernetes.io/docs/concepts/extend-kubernetes/compute-storage-net/network-plugins/
https://www.google.com/search?q=%22need+a+network+plugin%22+pod

(IMO) some addons provide network plugins; feasibly, one addon could give you a network plugin and something else at the same time (eg: aggregated API server).

Comment (aojea) [reply content/en/docs/concepts/cluster-administration/networking.md]: replaced "networking addon" by "network plugin"

File: content/en/docs/concepts/cluster-administration/networking.md
Comment (sftim) [inline content/en/docs/concepts/cluster-administration/networking.md:60]:
```suggestion
IP addresses on its interfaces, but only the IP addresses in a Node's `.status.addresses` or a Pod's `.status.ips` are
```

In the API , you won't find a field with a JSON path that looks like `node.status.addresses` or `pod.status.ips`. Nor the gRPC equivalent.

File: content/en/docs/concepts/cluster-administration/networking.md
Comment (sftim) [inline content/en/docs/concepts/cluster-administration/networking.md:58]:
```suggestion
Kubernetes clusters only consider the IP families present on the Pod, Service and Node objects,
```

File: content/en/docs/concepts/cluster-administration/networking.md
Comment (sftim) [inline content/en/docs/concepts/cluster-administration/networking.md:40]:
```suggestion
- The network plugin is responsible for assigning IP addresses to Pods.
```

---

Optionally:
```suggestion
- The network plugin is responsible for assigning IP addresses to Pods.
  Depending on your cluster, the IP addresses assigned to Pods could come from a private
  network range or from a block of public addresses (if you have one assigned).
```

File: content/en/docs/concepts/cluster-administration/networking.md
Comment (sftim) [inline content/en/docs/concepts/cluster-administration/networking.md:41]: Optionally:
```suggestion
- The kube-apiserver is configured to assign IP addresses to Services.
  Depending on your cluster, the IP addresses assigned to Services could come from a private
  network range or from a block of public addresses (if you have one assigned).
```

File: content/en/docs/concepts/cluster-administration/networking.md
Comment (sftim) [inline content/en/docs/concepts/cluster-administration/networking.md:42]:
```suggestion
- Either the kubelet or the cloud-controller-manager is configured to assign IP addresses to Nodes.
  In some clusters, IP address management for nodes happens outside of Kubernetes, and the
  control plane's only role is to record the IP address that has already been assigned.
  If you have a cluster autoscaler that sets up or terminates nodes to match demand, the autoscaler
  might use networking requirements to decide what to provision.
```

File: content/en/docs/concepts/cluster-administration/networking.md
Comment (sftim) [inline content/en/docs/concepts/cluster-administration/networking.md:44]:
```suggestion
{{< figure src="/docs/images/kubernetes-cluster-network.svg" alt="A figure illustrating the different network ranges in a Kubernetes cluster" class="diagram-medium" >}}
```

**Review by sftim — APPROVED** [review]
The feedback here is to record scope for improvement. However, I'm keen to see this change land as is.

Thanks for the PR!

/lgtm

**Pull request merged**
